service/tablet_allocator.cc

```diff
@@ -500,24 +502,31 @@ class load_balancer {
         std::vector<table_id_and_size_desc> tables_need_resize;
         std::vector<table_id_and_size_desc> tables_being_resized;
 
-        static bool table_needs_merge(const table_size_desc& d) {
-            // The initial_tablet_count is respected while the table is in "growing mode".
-            // We say that a table leaves this mode if it required a split above the initial
-            // tablet count. After that, we can rely purely on the average size to say that
-            // a table is shrinking and requires merge.
-            // FIXME: this is not perfect and we may want to leave the mode too if we detect
-            //  average size is decreasing significantly, before any split happened.
-            bool left_growing_mode = !d.resize_decision.initial_decision();
-            lblogger.debug("table_needs_merge: tablet_count={}, avg_tablet_size={}, left_growing_mode={} (seq number: {})",
-                           d.tablet_count, d.avg_tablet_size, left_growing_mode, d.resize_decision.sequence_number);
-            return left_growing_mode && d.tablet_count > 1 && d.avg_tablet_size < d.target_min_tablet_size();
-        }
-        static bool table_needs_split(const table_size_desc& d) {
-            return d.avg_tablet_size > d.target_max_tablet_size();
+        static locator::resize_decision to_resize_decision(const table_size_desc& d) {
+            locator::resize_decision decision;
+
+            auto tablet_count_from_size = d.tablet_count;
+            if (d.avg_tablet_size < d.target_min_tablet_size()) {
+                tablet_count_from_size /= 2;
+            } else if (d.avg_tablet_size > d.target_max_tablet_size()) {
+                tablet_count_from_size *= 2;
+            }
+
+            auto target_tablet_count = std::max(d.min_tablet_count, tablet_count_from_size);
+
```

Comment:
I guess the global decision that might want to bypass this limit in order to respect per-shard limit will be addressed by Tomek's work.

Reply:
it should. @tgrabiec?